### PR TITLE
feat: Allow users to comment

### PR DESCRIPTION
**Pull request by orronai**
- Users can now comment to their own solutions
- Users can now delete their own comments

### CODE REVIEW

File: lms/static/comments.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 const DEFAULT_COMMENTED_LINE_COLOR = '#fab3b0';
+const STUDENT_COMMENTED_LINE_COLOR = '#b2eefa';
```
Comment (yammesicka) [inline lms/static/comments.js:2]: Not such good selection for a color - can you try something a bit more dul? Maybe around #cdffd8. WDYT?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/templates/view.html [hunk 2]
```diff
@@ -131,8 +131,8 @@ <h2>{{ _('הערות בודק') }}</h2>
 <script src="{{ url_for('static', filename='prism.js') }}"></script>
 {% if not shared_url %}
 <script src="{{ url_for('static', filename='comments.js') }}"></script>
-{%- if is_manager %}
 <script src="{{ url_for('static', filename='grader.js') }}"></script>
+{%- if is_manager %}
```
Comment (yammesicka) [inline lms/templates/view.html:135]: This should be decided using settings

File: lms/lmsweb/views.py [hunk 3]
```diff
@@ -501,9 +501,13 @@ def _common_comments(exercise_id=None, user_id=None):
     Most common comments throughout all exercises.
     Filter by exercise id when specified.
     """
-    query = CommentText.filter(**{
-        CommentText.flake8_key.name: None,
-    }).select(CommentText.id, CommentText.text).join(Comment)
+    query = (
+        CommentText.select(CommentText.id, CommentText.text).join(Comment)
+        .join(User).join(Role).where(
+            CommentText.flake8_key.name is not None,
```
Comment (yammesicka) [inline lms/lmsweb/views.py:507]: According to [the documentation](http://docs.peewee-orm.com/en/latest/peewee/query_operators.html), this query might not work. Use `is_null` instead.

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/lmsweb/views.py [hunk 4]
```diff
@@ -274,6 +274,12 @@ def comment():
...
 
+    if (
+        not webapp.config.get('USERS_COMMENTS', False)
+        and not current_user.role.is_manager
+    ):
+        return fail(403, "You aren't allowed to access this page.")
+
     if act == 'delete':
         comment_id = int(request.args.get('commentId'))
         comment_ = Comment.get_or_none(Comment.id == comment_id)
```
Comment (yammesicka) [inline lms/lmsweb/views.py:285]: A user should be able to delete a comment only if its his comment , or he's an administrator

File: lms/lmsweb/views.py [hunk 5]
```diff
@@ -501,9 +507,13 @@ def _common_comments(exercise_id=None, user_id=None):
...
     Filter by exercise id when specified.
     """
-    query = CommentText.filter(**{
-        CommentText.flake8_key.name: None,
-    }).select(CommentText.id, CommentText.text).join(Comment)
+    query = (
+        CommentText.select(CommentText.id, CommentText.text).join(Comment)
+        .join(User).join(Role).where(
+            CommentText.flake8_key.is_null(True),
+            Comment.commenter.role > Role.get_student_role().id,
```
Comment (yammesicka) [inline lms/lmsweb/views.py:514]: `role` doesn't have total ordering, and we don't wanna take the comments of banned users.

Check if role == administrator or role == staff

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/lmsweb/views.py [hunk 6]
```diff
@@ -501,9 +512,17 @@ def _common_comments(exercise_id=None, user_id=None):
     Most common comments throughout all exercises.
     Filter by exercise id when specified.
     """
-    query = CommentText.filter(**{
-        CommentText.flake8_key.name: None,
-    }).select(CommentText.id, CommentText.text).join(Comment)
+    is_moderator_comments = (
+        (Comment.commenter.role == Role.get_staff_role().id)
```
Comment (yammesicka) [inline lms/lmsweb/views.py:516]: `Comment.commenter.role.is_staff`?

**Review by yammesicka — APPROVED** [review]

File: lms/lmsweb/views.py (hunk 6 above)
Comment (orronai) [reply lms/lmsweb/views.py:516]: It's not possible - it throws an error that says is_staff, is_manager etc. are not attributes of Role.

**Pull request merged**